### PR TITLE
correct version check when we have a dev version installed

### LINKED ISSUE #14641: Let pulumi CLI check for newer dev versions
We're introducing a dev release channel in https://github.com/pulumi/pulumi/issues/14622.  The pulumi CLI currently warns about outdated versions.  We should be able to warn about new dev versions as well when they are available, and the user is on a dev release, so they can always update to the latest version.  This depends at least on https://github.com/pulumi/pulumi/issues/14623 as well.

/xref https://github.com/pulumi/home/issues/3003

### PR DESCRIPTION
**Pull request by tgummerer**
When we have a dev version from the dev CLI channel installed, we want to warn whenever a new version is available, so the users can upgrade. Do that, and also reduce the amount of time the version information is cached, since we expect dev versions to be released much more frequently than regular releases.

This needs https://github.com/pulumi/pulumi-service/pull/17429 to be merged and deployed first.

Fixes #14641

## Checklist

- [ ] I have run `make tidy` to update any new dependencies
- [ ] I have run `make lint` to verify my code passes the lint check
  - [ ] I have formatted my code using `gofumpt`

- [ ] I have added tests that prove my fix is effective or that my feature works

- [ ] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

File: pkg/backend/httpstate/client/client_test.go [hunk 1]
```diff
@@ -105,12 +105,31 @@ func TestAPIErrorResponses(t *testing.T) {
...
+		_, _, _, defaultErrorErr := defaultErrorClient.GetCLIVersionInfo(context.Background())
 
 		assert.Error(t, defaultErrorErr)
 	})
 }
 
+func TestAPIVersionResponses(t *testing.T) {
+	t.Parallel()
+
+	// test handling non-standard error message
```
Comment (Frassle) [inline pkg/backend/httpstate/client/client_test.go:117]: Odd comment?

Comment (tgummerer) [reply pkg/backend/httpstate/client/client_test.go]: I was trusting copilot a bit much :grimacing: fixed.

File: pkg/cmd/pulumi/pulumi.go [hunk 2]
```diff
@@ -397,6 +398,38 @@ func NewPulumiCmd() *cobra.Command {
 	return cmd
 }
 
+// haveNewerDevVersion checks whethere we have a newer dev version available.
+func haveNewerDevVersion(devVersion semver.Version, curVersion semver.Version) bool {
```
Comment (tgummerer) [inline pkg/cmd/pulumi/pulumi.go:402]: This function is intentionally lenient with any errors, erring on the side of pretending we don't have a newer version available.  I expect this to be a better user experience than adding any asserts here for correctness, as this affects merely a warning, and if we don't end up showing it it should not be a big problem.

File: pkg/cmd/pulumi/pulumi.go [hunk 3]
```diff
@@ -397,6 +398,38 @@ func NewPulumiCmd() *cobra.Command {
...
+	}
+	if devVersion.Patch != curVersion.Patch {
+		return devVersion.Patch > curVersion.Patch
+	}
+
+	// The dev version string looks like: v1.0.0-11-g4ff08363.  We
+	// can determine whether we have a newer dev version by
+	// comparing the second part of the version string, which is
+	// the number of commits since the last tag.
+	devVersionParts := strings.Split(devVersion.String(), "-")
```
[the remark this answers is not in the document]
Comment (tgummerer) [reply pkg/cmd/pulumi/pulumi.go:417]: No, the `Pre` array is not split by `-`, it's split by `.`s. So we could look at `Pre[0]`, but we'd still need string splitting (`Pre[0]` looks something like `11-gdeadbeef`, and we'd also need to check that `Pre` is not an empty array (it never should be, but better safe than sorry here I think). So this seemed like the simpler thing to do.

File: pkg/cmd/pulumi/pulumi.go (hunk 2 above)
Comment (tgummerer) [reply pkg/cmd/pulumi/pulumi.go:402]: Good idea, done :heavy_check_mark:

**Review by julienp — APPROVED** [review]
:shipit:

**Pull request merged**